pageserver/src/restore_local_repo.rs

```diff
@@ -426,14 +425,38 @@ pub fn save_decoded_record(
             relnode: blk.rnode_relnode,
             forknum: blk.forknum as u8,
         });
-
-        let rec = WALRecord {
-            will_init: blk.will_init || blk.apply_image,
-            rec: recdata.clone(),
-            main_data_offset: decoded.main_data_offset as u32,
-        };
-
-        timeline.put_wal_record(lsn, tag, blk.blkno, rec)?;
+        if blk.apply_image
+            && blk.has_image
+            && decoded.xl_rmid == pg_constants::RM_XLOG_ID
+            && (decoded.xl_info == pg_constants::XLOG_FPI
+                || decoded.xl_info == pg_constants::XLOG_FPI_FOR_HINT)
+        {
+            // Extract page image from FPI record
+            let img_len = blk.bimg_len as usize;
+            let img_offs = blk.bimg_offset as usize;
+            let mut image = BytesMut::with_capacity(pg_constants::BLCKSZ as usize);
+            image.extend_from_slice(&recdata[img_offs..img_offs + img_len]);
+
+            // Compression of WAL is not yet supported
+            assert!((blk.bimg_info & pg_constants::BKPIMAGE_IS_COMPRESSED) == 0);
+
+            if blk.hole_length != 0 {
+                let tail = image.split_off(blk.hole_offset as usize);
+                image.resize(image.len() + blk.hole_length as usize, 0u8);
```

Comment:
Does this work with checksums enabled?

Reply:
It is interesting question ... having no relation to this patch!
Actually checksums are not working with our current Zenith architecture.
Checksums are calculated when page is written to the file from shared buffers (FlushBuffer).
But we are not writing anything to the file at compute node. And neither we do it at wal-redo Postgres, where page image is just copied from shared buffer to the pipe. In principle it is possible to call PageSetChecksumCopy here.
But it is not done.

Calling of PageSetChecksumCopy from Rust code converting FPI record to pageimage is a little bit more problematic.
But also can be done.

In any case, checksums are needed to verify that page is not corrupted on the disk.
Right now pageserver is storing pages in its own way: layered repo as part of image layers, buffered repo: as scattered compressed records.  As far as compression calculates its own checksum, there is no need to use Postgres checksums.

Reply:
Wow, I did not realize that we didn't support checksums... 

Do we want / need to add that to the TODO list? Or is this a known limitation / advertised as such in Console?